### PR TITLE
allow template calls in initializers, rewrite template_call_innards

### PR DESCRIPTION
**Pull request by matter123**
Fixes: #307
Fixes: #314

### CODE REVIEW

**Comment by jeff-hykin** [comment]
I added subroutines to the library to enable the templates to be matched to any depth. I also took away the "remove_numbered_capture_groups" so that the inherited type can be fully matched. Once I merge in retagging we can add the "inherited" back to it

**Comment by jeff-hykin** [comment]
After you've looked over it we can merge this into master

File: source/languages/cpp/generate.rb [hunk 1]
```diff
@@ -724,27 +724,45 @@ def generateBlockFinder( name:"", tag_as:"", start_pattern:nil, needs_semicolon:
...
-    no_brackets_at_all = /[^<>]*/
-    balanced_brackets = /[^>]*+<[^>]*+>/
-    some_number_of_angle_brackets = oneOrMoreOf(no_brackets_at_all.or(balanced_brackets))
-    # TODO: change this to a readble form (above) once possessives (aka no back_track) is imlpemented
-    # otherwise this pattern fails because its to computationally expensive
-    some_number_of_angle_brackets = /(?:[^<>]++(?:<[^<>]++>)?)++/
+    no_anglebrackets_at_all = /[^<>]*/
+    balanced_brackets = //
+    # this is effectively what is happening (below)
+    some_number_of_angle_brackets = oneOrMoreOf(no_anglebrackets_at_all.or(balanced_brackets))
```
Comment (matter123) [inline source/languages/cpp/generate.rb]: This should probably be a comment.

File: source/languages/cpp/generate.rb [hunk 2]
```diff
@@ -3768,35 +3785,6 @@ def generateBlockFinder( name:"", tag_as:"", start_pattern:nil, needs_semicolon:
                 include: "#preprocessor_rule_define_line_context"
             }
         ]
-    legacy_memory_new_call = {
```
Comment (matter123) [inline source/languages/cpp/generate.rb]: Why was this removed?

Comment (jeff-hykin) [reply source/languages/cpp/generate.rb]: it wasn't used anywhere and I just realized that when I was trying to remove the `template_call.without_numbered_capture_groups`

**Comment by jeff-hykin** [comment]
> Changes look good. The only slight issue is that in isolation it is not immediately clear that `rematch` performs a subexp call rather than a back reference. But I can't thnk of a better name.

Yeah I agree. I named it `recurisvelyMatch()` the first time, but its not always recursive. Then again, if its not recursive its pretty easy to just copy and paste the pattern so maybe it should be named `recurisvelyMatch()`

**Comment by jeff-hykin** [comment]
`recursivelyMatch()` and `matchResultOf()` are still not great names so let me know if you think of anything better
